libsql-wal/src/bins/compactor/main.rs

```diff
@@ -0,0 +1,225 @@
+use std::path::PathBuf;
+
+use anyhow::Context;
+use aws_config::{retry::RetryConfig, BehaviorVersion, Region};
+use aws_credential_types::Credentials;
+use aws_sdk_s3::config::SharedCredentialsProvider;
+use clap::{Parser, ValueEnum};
+use libsql_wal::io::StdIO;
+use libsql_wal::storage::backend::s3::S3Backend;
+use libsql_wal::storage::compaction::strategy::{
+    identity::IdentityStrategy, log_strategy::LogReductionStrategy, PartitionStrategy,
+};
+use libsql_wal::storage::compaction::Compactor;
+
+#[derive(Debug, clap::Args)]
+struct S3Args {
+    #[arg(long, requires = "S3Args")]
+    enable_s3: bool,
+    #[arg(long, env = "LIBSQL_BOTTOMLESS_DATABASE_ID")]
+    cluster_id: Option<String>,
+    #[arg(long, env = "LIBSQL_BOTTOMLESS_ENDPOINT")]
+    s3_url: Option<String>,
+    #[arg(long, env = "LIBSQL_BOTTOMLESS_AWS_SECRET_ACCESS_KEY")]
+    s3_access_key: Option<String>,
+    #[arg(long, env = "LIBSQL_BOTTOMLESS_AWS_ACCESS_KEY_ID")]
+    s3_access_key_id: Option<String>,
+    #[arg(long, env = "LIBSQL_BOTTOMLESS_BUCKET")]
+    s3_bucket: Option<String>,
+    #[arg(long, env = "LIBSQL_BOTTOMLESS_AWS_DEFAULT_REGION")]
+    s3_region_id: Option<String>,
+}
+
+#[derive(Clone, Debug, ValueEnum)]
+enum CompactStrategy {
+    Logarithmic,
+    CompactAll,
+}
+
+#[derive(Debug, clap::Parser)]
+struct Command {
+    #[arg(long, short, default_value = "compactor")]
+    path: PathBuf,
+    #[command(flatten)]
+    s3_args: S3Args,
+    #[command(subcommand)]
+    subcommand: Subcommand,
+}
+
+#[derive(Debug, clap::Subcommand)]
+enum Subcommand {
+    /// Register namespaces to monitor
+    Monitor { namespace: String },
+    /// Analyze segments for a namespaces
+    Analyze {
+        /// list all segments
+        #[clap(long)]
+        list_all: bool,
+        namespace: String,
+    },
+    /// Compact segments into bigger segments
+    Compact {
+        // compaction strategy
+        #[clap(long, short)]
+        strategy: CompactStrategy,
+        /// prints the compaction plan, but doesn't perform it.
+        #[clap(long)]
+        dry_run: bool,
+        namespace: String,
+    },
+    /// Sync namespace metadata from remote storage
+    Sync {
+        /// When performing a full sync, all the segment space is scanned again. By default, only
+        /// segments with frame_no greated that the last frame_no are retrieved.
+        #[clap(long)]
+        full: bool,
+        /// unless this is specified, all monitored namespaces are synced
+        namespace: Option<String>,
+    },
+    /// Restore namespace
+    Restore {
+        #[clap(long)]
+        verify: bool,
+        namespace: String,
+        out: PathBuf,
+    },
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    let cmd: Command = Command::parse();
+
+    let backend = setup_storage(&cmd.s3_args).await?;
+    tokio::fs::create_dir_all(&cmd.path).await?;
```

Comment:
Should we check if the dir already exists and refuse to work if it does to prevent mistakes of messing up with some existing files?

Reply:
`create_dir_all` only creates if the dir doesn't already exist, if that's your concern